include/unifex/let_error.hpp

```diff
@@ -33,89 +34,96 @@
+template <typename Source, typename Func, typename Receiver>
 struct _rcvr {
   class type;
 };
-template <typename Source, typename Error, typename Receiver>
-using receiver_type = typename _rcvr<Source, Error, remove_cvref_t<Receiver>>::type;
+template <typename Source, typename Func, typename Receiver>
+using receiver_type = typename _rcvr<Source, Func, remove_cvref_t<Receiver>>::type;
 
-template <typename Source, typename Error, typename Receiver>
+template <typename Source, typename Func, typename Receiver, typename Error>
 struct _frcvr {
   class type;
 };
-template <typename Source, typename Error, typename Receiver>
-using final_receiver_type = typename _frcvr<Source, Error, remove_cvref_t<Receiver>>::type;
+template <typename Source, typename Func, typename Receiver, typename Error>
+using final_receiver_type =
+    typename _frcvr<Source, Func, remove_cvref_t<Receiver>, Error>::type;
 
-template <typename Source, typename Error>
+template <typename Source, typename Func>
 struct _sndr {
   class type;
 };
 
-template <typename Source, typename Error>
-using _sender = typename _sndr<Source, Error>::type;
+template <typename Source, typename Func>
+using _sender = typename _sndr<Source, Func>::type;
 
-enum class state { neither, source, final };
-
-template <typename Source, typename Error, typename Receiver>
-class _rcvr<Source, Error, Receiver>::type final {
-  using operation = operation_type<Source, Error, Receiver>;
-  using final_receiver = final_receiver_type<Source, Error, Receiver>;
-  using final_sender_t = callable_result_t<Error&>;
+template <typename Source, typename Func, typename Receiver>
+class _rcvr<Source, Func, Receiver>::type final {
+  using operation = operation_type<Source, Func, Receiver>;
+  template <typename Error>
+  using final_receiver = final_receiver_type<Source, Func, Receiver, Error>;
 
 public:
-  explicit type(operation* op) noexcept
-    : op_(op) {}
-
-  type(type&& other) noexcept
-    : op_(std::exchange(other.op_, {}))
-  {}
- 
+  explicit type(operation* op) noexcept : op_(op) {}
+  type(type&& other) noexcept : op_(std::exchange(other.op_, {})) {}
+
+  // Taking by value here to force a move/copy on the offchange the value
+  // objects live in the operation state, in which case destroying the
+  // predecessor operation state would invalidate it.
+  //
+  // Flipping the order of `deactivate_*` and `set_value` is UB since by the
+  // time `set_value()` returns `op_` might as well be already destroyed,
+  // without proper deactivation of `op_->sourceOp_`.
   template(typename... Values)
     (requires receiver_of<Receiver, Values...>)
-  void set_value(Values&&... values) noexcept(
+  void set_value(Values... values) noexcept(
       is_nothrow_receiver_of_v<Receiver, Values...>) {
     UNIFEX_ASSERT(op_ != nullptr);
+    unifex::deactivate_union_member(op_->sourceOp_);
     unifex::set_value(std::move(op_->receiver_), (Values&&)values...);
```

Comment:
I wish this said `std::move(values)...`; this is a `move` case, not a forward case.

Reply:
Addressed in #450